File: ios/WorkoutRecord/WorkoutRecord/Sources/Models/UserGoal.swift
import Foundation

// MARK: - User Goal

struct UserGoal: Identifiable, Codable {
    let id: UUID
    let userId: UUID
    var weeklyWorkoutGoal: Int           // 週訓練次數目標
    var targetWeight: Double?             // 目標體重
    var volumeGoals: VolumeGoals          // 各肌群容量目標
    var restDayReminder: Bool             // 休息日提醒
    let createdAt: Date
    var updatedAt: Date
    
    init(
        id: UUID = UUID(),
        userId: UUID,
        weeklyWorkoutGoal: Int = 3,
        targetWeight: Double? = nil,
        volumeGoals: VolumeGoals = VolumeGoals(),
        restDayReminder: Bool = false,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.userId = userId
        self.weeklyWorkoutGoal = weeklyWorkoutGoal
        self.targetWeight = targetWeight
        self.volumeGoals = volumeGoals
        self.restDayReminder = restDayReminder
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

// MARK: - Volume Goals

struct VolumeGoals: Codable {
    var chest: Double?      // 胸部週容量目標 (kg)
    var back: Double?       // 背部週容量目標
    var legs: Double?       // 腿部週容量目標
    var shoulders: Double?  // 肩部週容量目標
    var arms: Double?       // 手臂週容量目標
    var core: Double?       // 核心週容量目標
    
    init(
        chest: Double? = nil,
        back: Double? = nil,
        legs: Double? = nil,
        shoulders: Double? = nil,
        arms: Double? = nil,
        core: Double? = nil
    ) {
        self.chest = chest
        self.back = back
        self.legs = legs
        self.shoulders = shoulders
        self.arms = arms
        self.core = core
    }
    
    // 根據肌群獲取目標
    func goal(for muscleGroup: Exercise.PrimaryMuscleGroup) -> Double? {
        switch muscleGroup {
        case .chest: return chest
        case .back: return back
        case .legs: return legs
        case .shoulders: return shoulders
        case .arms: return arms
        case .core: return core
        @unknown default: return nil
        }
    }
    
    // 設定肌群目標
    mutating func setGoal(for muscleGroup: Exercise.PrimaryMuscleGroup, value: Double?) {
        switch muscleGroup {
        case .chest: chest = value
        case .back: back = value
        case .legs: legs = value
        case .shoulders: shoulders = value
        case .arms: arms = value
        case .core: core = value
        @unknown default: break
        }
    }
}

// MARK: - Goal Progress

struct GoalProgress {
    let current: Double
    let target: Double
    
    var percentage: Double {
        guard target > 0 else { return 0 }
        return (current / target) * 100
    }
    
    var isComplete: Bool {
        percentage >= 100
    }
    
    var remaining: Double {
        max(0, target - current)
    }
}

File: app/lib/data/models/user_goal.dart
import 'package:drift/drift.dart' show Value;

import '../db/app_database.dart' as db;
import 'exercise.dart';

/// 各肌群的週容量目標(kg)。對應 Drift `UserGoals` 表中的 `*VolumeGoal` 欄位。
class VolumeGoals {
  final double? chest;
  final double? back;
  final double? legs;
  final double? shoulders;
  final double? arms;
  final double? core;

  const VolumeGoals({
    this.chest,
    this.back,
    this.legs,
    this.shoulders,
    this.arms,
    this.core,
  });

  /// 依肌群取得目標值。
  double? goalFor(PrimaryMuscleGroup muscleGroup) {
    switch (muscleGroup) {
      case PrimaryMuscleGroup.chest:
        return chest;
      case PrimaryMuscleGroup.back:
        return back;
      case PrimaryMuscleGroup.legs:
        return legs;
      case PrimaryMuscleGroup.shoulders:
        return shoulders;
      case PrimaryMuscleGroup.arms:
        return arms;
      case PrimaryMuscleGroup.core:
        return core;
      case PrimaryMuscleGroup.glutes:
      case PrimaryMuscleGroup.fullBody:
        return null;
    }
  }

  VolumeGoals copyWith({
    double? chest,
    double? back,
    double? legs,
    double? shoulders,
    double? arms,
    double? core,
  }) {
    return VolumeGoals(
      chest: chest ?? this.chest,
      back: back ?? this.back,
      legs: legs ?? this.legs,
      shoulders: shoulders ?? this.shoulders,
      arms: arms ?? this.arms,
      core: core ?? this.core,
    );
  }
}

/// 使用者訓練目標,對應 Drift `UserGoals` 表(每位使用者一筆)。
class UserGoal {
  final String id;
  final String userId;
  final int weeklyWorkoutGoal;
  final double? targetWeight;
  final VolumeGoals volumeGoals;
  final bool restDayReminder;
  final DateTime createdAt;
  final DateTime updatedAt;

  const UserGoal({
    required this.id,
    required this.userId,
    this.weeklyWorkoutGoal = 3,
    this.targetWeight,
    this.volumeGoals = const VolumeGoals(),
    this.restDayReminder = false,
    required this.createdAt,
    required this.updatedAt,
  });

  factory UserGoal.fromRow(db.UserGoal row) {
    return UserGoal(
      id: row.id,
      userId: row.userId,
      weeklyWorkoutGoal: row.weeklyWorkoutGoal,
      targetWeight: row.targetWeight,
      volumeGoals: VolumeGoals(
        chest: row.chestVolumeGoal,
        back: row.backVolumeGoal,
        legs: row.legsVolumeGoal,
        shoulders: row.shouldersVolumeGoal,
        arms: row.armsVolumeGoal,
        core: row.coreVolumeGoal,
      ),
      restDayReminder: row.restDayReminder,
      createdAt: row.createdAt,
      updatedAt: row.updatedAt,
    );
  }

  db.UserGoalsCompanion toCompanion() {
    return db.UserGoalsCompanion(
      id: Value(id),
      userId: Value(userId),
      weeklyWorkoutGoal: Value(weeklyWorkoutGoal),
      targetWeight: Value(targetWeight),
      chestVolumeGoal: Value(volumeGoals.chest),
      backVolumeGoal: Value(volumeGoals.back),
      legsVolumeGoal: Value(volumeGoals.legs),
      shouldersVolumeGoal: Value(volumeGoals.shoulders),
      armsVolumeGoal: Value(volumeGoals.arms),
      coreVolumeGoal: Value(volumeGoals.core),
      restDayReminder: Value(restDayReminder),
      createdAt: Value(createdAt),
      updatedAt: Value(updatedAt),
    );
  }

  UserGoal copyWith({
    int? weeklyWorkoutGoal,
    double? targetWeight,
    VolumeGoals? volumeGoals,
    bool? restDayReminder,
    DateTime? updatedAt,
  }) {
    return UserGoal(
      id: id,
      userId: userId,
      weeklyWorkoutGoal: weeklyWorkoutGoal ?? this.weeklyWorkoutGoal,
      targetWeight: targetWeight ?? this.targetWeight,
      volumeGoals: volumeGoals ?? this.volumeGoals,
      restDayReminder: restDayReminder ?? this.restDayReminder,
      createdAt: createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
